radix-engine/src/engine/heap.rs

```diff
@@ -112,3 +149,63 @@ impl Heap {
             .ok_or(CallFrameError::RENodeNotOwned(node_id))
     }
 }
+
+#[derive(Debug)]
+pub struct HeapRENode {
+    pub substates: HashMap<SubstateOffset, RuntimeSubstate>,
+    pub child_nodes: HashSet<RENodeId>,
+}
+
+impl HeapRENode {
+    // TODO: Move this into kernel
+    pub fn try_drop(&mut self) -> Result<(), DropFailure> {
+        // TODO: Remove this
+        if !self.child_nodes.is_empty() {
+            return Err(DropFailure::DroppingNodeWithChildren);
+        }
+
+        for (_, substate) in &mut self.substates {
+            match substate {
+                RuntimeSubstate::AuthZone(auth_zone) => {
+                    auth_zone.clear_all();
+                    Ok(())
+                }
+                RuntimeSubstate::GlobalRENode(..) => panic!("Should never get here"),
+                RuntimeSubstate::Package(..) => Err(DropFailure::Package),
+                RuntimeSubstate::Vault(..) => Err(DropFailure::Vault),
+                RuntimeSubstate::KeyValueStoreEntry(..) => Err(DropFailure::KeyValueStore),
+                RuntimeSubstate::NonFungible(..) => Err(DropFailure::NonFungibleStore),
+                RuntimeSubstate::ComponentInfo(..) => Err(DropFailure::Component),
+                RuntimeSubstate::ComponentState(..) => Err(DropFailure::Component),
+                RuntimeSubstate::Bucket(..) => Ok(()),
+                RuntimeSubstate::ResourceManager(..) => Err(DropFailure::Resource),
+                RuntimeSubstate::System(..) => Err(DropFailure::System),
+                RuntimeSubstate::Proof(proof) => {
+                    proof.drop();
+                    Ok(())
+                }
+                RuntimeSubstate::Worktop(worktop) => worktop.drop(),
+            }?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Into<BucketSubstate> for HeapRENode {
+    fn into(mut self) -> BucketSubstate {
+        self.substates
+            .remove(&SubstateOffset::Bucket(BucketOffset::Bucket))
+            .unwrap()
+            .into()
+    }
+}
+
+impl Into<ProofSubstate> for HeapRENode {
+    fn into(mut self) -> ProofSubstate {
+        self.substates
+            .remove(&SubstateOffset::Proof(ProofOffset::Proof))
+            .unwrap()
+            .into()
+    }
+}
```

Comment:
This is a general comment. I think we should use `TryInto` trait for conversions that may fail.